Fix TkAgg memory leaks and test for memory growth regressions

richardsheridan opened the pull request:
## PR Summary
This PR is an attempt to fix #20490. Basically it allows a normally-forbidden `self.window.update()` if we detect that the `tkinter` mainloop is not running. The new test currently still fails though because of some additional leak related to the HiDPI stuff in #19167.

I took the liberty of introducing a new dependency in the test suite. Let me know if I should insert it into any other files or if you would rather find a different way to observe the leak.
## PR Checklist

<!-- Please mark any checkboxes that do not apply to this PR as [N/A]. -->
**Tests and Styling**
- [x] Has pytest style unit tests
- [ ]  (and `pytest` passes).
- [x] Is [Flake 8](https://flake8.pycqa.org/en/latest/) compliant (install `flake8-docstrings` and run `flake8 --docstring-convention=all`).

**Documentation**
- [ ] ~~New features are documented, with examples if plot related.~~
- [ ] ~~New features have an entry in `doc/users/next_whats_new/` (follow instructions in README.rst there).~~
- [ ] ~~API changes documented in `doc/api/next_api_changes/` (follow instructions in README.rst there).~~
- [ ] ~~Documentation is sphinx and numpydoc compliant (the docs should [build](https://matplotlib.org/devel/documenting_mpl.html#building-the-docs) without error).~~

<!--
Thank you so much for your PR!  To help us review your contribution, please
consider the following points:

- A development guide is available at https://matplotlib.org/devdocs/devel/index.html.

- Help with git and github is available at
  https://matplotlib.org/devel/gitwash/development_workflow.html.

- Do not create the PR out of main, but out of a separate branch.

- The PR title should summarize the changes, for example "Raise ValueError on
  non-numeric input to set_xlim".  Avoid non-descriptive titles such as
  "Addresses issue #8576".

- The summary should provide at least 1-2 sentences describing the pull request
  in detail (Why is this change required?  What problem does it solve?) and
  link to any relevant issues.

- If you are contributing fixes to docstrings, please pay attention to
  http://matplotlib.org/devel/documenting_mpl.html#formatting.  In particular,
  note the difference between using single backquotes, double backquotes, and
  asterisks in the markup.

We understand that PRs can sometimes be overwhelming, especially as the
reviews start coming in.  Please let us know if the reviews are unclear or
the recommended next step seems overly demanding, if you would like help in
addressing a reviewer's comments, or if you have been waiting too long to hear
back on your PR.
-->

tacaswell commented on lib/matplotlib/tests/test_backends_interactive.py:
```diff
@@ -319,3 +318,46 @@ def test_lazy_linux_headless():
+
+@pytest.mark.parametrize("env", _get_testable_interactive_backends())
+def test_figure_leak_20490(env):
+    subprocess.run(
```
Maybe make this a skip if psutil is not installed?

tacaswell commented:
Hopefully adding a test dependency in a patch release will not be too troublesome for the packagers.  Assume the escape hatch to patch out the test.

tacaswell commented:
@richardsheridan Thank you for tracking this down and 👍🏻 on adding that test!

richardsheridan commented:
@tacaswell should I XFAIL the test for now or just let it fail entirely?

tacaswell commented:
Can you uses `pytest.importorskip` https://docs.pytest.org/en/6.2.x/reference.html#pytest-importorskip ?  We already use it in the tests to skip the tests that require pandas.

QuLogic commented on lib/matplotlib/tests/test_backends_interactive.py:
```diff
@@ -319,3 +318,47 @@ def test_lazy_linux_headless():
+        [sys.executable, "-c",
+         inspect.getsource(_test_figure_leak) + "\n_test_figure_leak()"],
+        env={**os.environ, "SOURCE_DATE_EPOCH": "0", **env},
+        timeout=_test_timeout, check=True,)
```
```suggestion
        timeout=_test_timeout, check=True)
```

QuLogic commented on lib/matplotlib/tests/test_backends_interactive.py:
```diff
@@ -319,3 +318,47 @@ def test_lazy_linux_headless():
+    fig = plt.figure()
+    plt.close(fig)
+    growth = p.memory_full_info().uss - mem
+    assert growth < ACCEPTABLE_MEMORY_LEAKAGE, "Memory leakage too high"
```
Since this is failing, probably want to mention by how much (this isn't run in pytest, so won't get any argument parsing).

richardsheridan commented on lib/matplotlib/tests/test_backend_tk.py:
```diff
@@ -40,14 +40,14 @@ def test_func():
                 timeout=_test_timeout,
                 stdout=subprocess.PIPE,
                 stderr=subprocess.PIPE,
-                check=True,
```
In order to print the stderr below I could either inspect the `CalledProcessError` or the `CompletedProcess` and it seemed simpler just to check the return code manually. It also avoids some exceptions-as-control-flow and unnecessary exception chaining for a more readable traceback.

richardsheridan commented:
This PR is blocked by the hidpi leak that maybe only @QuLogic can debug. Otherwise we could adjust the memory leak threshold higher so it passes for now. That way the hidpi leak can be fixed later and th threshold can be lowered again at that time. Thoughts?

QuLogic commented:
I don't have Windows to check the HiDPI code right now. But I'm not sure how it could be leaking? The only HiDPI variable is attached to the window (the figure manager does `window_dpi = tk.IntVar(master=window, ...)` and `self._window_dpi = window_dpi`), so it should go away with it.

QuLogic commented:
I suppose there's also the additional `Font` object on the toolbar as well, but adding an explicit `del self.toolbar._label_font` to `delayed_destroy` doesn't do anything.

tacaswell commented:
If I run `memleak.py` with the following patch:

```diff
$ git diff
diff --git a/tools/memleak.py b/tools/memleak.py
index 9b9da912b7..d61d0b57c2 100755
--- a/tools/memleak.py
+++ b/tools/memleak.py
@@ -79,10 +79,10 @@ def run_memleak_test(bench, iterations, report):
     ax3.plot(open_files_arr)
     ax3.set_ylabel('open file handles')
 
-    if not report.endswith('.pdf'):
-        report = report + '.pdf'
+    if not report.endswith('.png'):
+        report = report + '.png'
     fig.tight_layout()
-    fig.savefig(report, format='pdf')
+    fig.savefig(report, format='png')
 
 
 class MemleakTest:
@@ -115,7 +115,7 @@ class MemleakTest:
             ax.pcolor(10 * np.random.rand(50, 50))
 
         fig.savefig(BytesIO(), dpi=75)
-        fig.canvas.flush_events()
+        # fig.canvas.flush_events()
         plt.close(1)

```

## Main branch 

![tkagg_main_no_flush](https://user-images.githubusercontent.com/199813/162080687-7ced8c2a-d02f-403a-b4e4-847c1c7c7587.png)


## this branch

![tkagg_branch_no_flush](https://user-images.githubusercontent.com/199813/162080834-ae17f411-19ab-44c9-b557-fb40bab91a73.png)

----

This clearly shows that this branch fixes something, but that the warm up time is in the hundreds of figures which is not practical for the test suite.

QuLogic commented on lib/matplotlib/tests/test_backends_interactive.py:
```diff
@@ -503,3 +502,46 @@ def test_blitting_events(env):
+        fig = plt.figure()
+        if t:
+            plt.pause(t)
+        plt.close(fig)
```
In a standalone script, it appears that if you add a `gc.collect` at the end of the loop, that will fix Tk with 0.0 time. Strangely, adding a _single_ `gc.collect` at the end of everything doesn't free up everything.

For some reason, this does not work in the test, but it does cut the counted growth in half, and is definitely something we should add to the test.

QuLogic replied:
I added this.

QuLogic commented:
I took the test and put it in a standalone script, and ran it through scalene. Unfortunately, it seems to break Tk threading with explicit `gc.collect` calls. With garbage collection, it seemed to think that `FigureCanvasTk._tkphoto` and `FigureCanvasTk._tkcanvas.create_image` were the biggest leak. However, I don't know how much we can trust that without garbage collection.

That being said, @tacaswell noticed that we explicitly deleted the image created from the photo on the canvas during resize:
https://github.com/matplotlib/matplotlib/blob/0e46ff26afa85d069d72b5ff3c97a040ee8b486c/lib/matplotlib/backends/_backend_tk.py#L235-L239

So I wonder if we're somehow missing deletion of that image when destroying the figure?

richardsheridan commented:
I made a bit of progress on this thanks to learning about `memleak.py`. The ever-increasing not-garbage objects before this PR was due to tkinter hanging on to various callbacks. I identified them using [`objgraph.show_growth`](https://mg.pov.lt/objgraph/#memory-leak-example) and following [`objgraph.show_backrefs`](https://mg.pov.lt/objgraph/#backreferences) from them. Confusingly they lead to nowhere, and they also show up with [`objgraph.get_leaking_objects`](https://mg.pov.lt/objgraph/objgraph.html#objgraph.get_leaking_objects). They aren't actually leaked, though... they are being legitimately [tracked](https://github.com/python/cpython/blob/c2e3c06139e9468efb32629d147d99a1672d9e19/Modules/_tkinter.c#L2325-L2326) by the tkinter c module. these are _mostly_ cleared via running the event loop in the patch.

I say mostly, but actually all were gone but one; the `CallWrapper` associated with [`FigureManagerTk._update_window_dpi`](https://github.com/matplotlib/matplotlib/blob/a1f2e5ff919031bdb4d2aa272459d6dc5b474830/lib/matplotlib/backends/_backend_tk.py#L434-L439).  This `CallWrapper` is created by the `trace_add` command which, much to my surprise, _globally_ registers the callback, and does not die with the [`IntVar`](https://github.com/matplotlib/matplotlib/blob/a1f2e5ff919031bdb4d2aa272459d6dc5b474830/lib/matplotlib/backends/_backend_tk.py#L426-L427) either on `destroy` or on normal python GC. So the entire web of figures and backend python objects is being pinned alive by that callback.

I manually unregister the trace in [a3016af](https://github.com/matplotlib/matplotlib/pull/22002/commits/a3016af42824bce342282abbab82222cd4a27b7c), and accellerate the garbage collection in [e6e4cad](https://github.com/matplotlib/matplotlib/pull/22002/commits/e6e4cad54d1a9079e47b92a7b4193626440373ae). This gave me flat graphs for memory and object behavior in `memleak.py`. Yay! But tkagg still fails my new test. Boo! 

I updated the test to flush events after both creating and closing the figure and adjusted the thresholds. It passes on my machine and we'll see if CI accepts it.

richardsheridan commented:
If I change `memleak.py` a bit more:
```diff
$  git diff
diff --git a/tools/memleak.py b/tools/memleak.py
index 9b9da912b7..6dbc9dc505 100755
--- a/tools/memleak.py
+++ b/tools/memleak.py

@@ -115,8 +121,9 @@ class MemleakTest:
             ax.pcolor(10 * np.random.rand(50, 50))

         fig.savefig(BytesIO(), dpi=75)
-        fig.canvas.flush_events()
+       plt.pause(.1)
         plt.close(1)


 if __name__ == '__main__':
```

I see continued growth in RSS in the tkagg backend, but constant object count which is disturbing.

![memleakpy](https://user-images.githubusercontent.com/8000454/162356881-b3f24e50-9505-4a7a-8553-2346ab9d2eb3.png)

Nevertheless, the growth is muted compared to before the patches.

richardsheridan commented:
I cannot reproduce the error in CI that has to do with a missing `update` command locally, but it must have something to do with the difference between `flush_events` and `pause` after the figure is destroyed.  Pause seems more leaky than flush_events so I reverted my earlier tweak, but the test will still fail on tkagg at the moment.

QuLogic commented on lib/matplotlib/tests/test_backends_interactive.py:
```diff
@@ -195,8 +195,7 @@ def _test_thread_impl():
     future.result()  # Joins the thread; rethrows any exception.
-    plt.close()
-    fig.canvas.flush_events()  # pause doesn't process events after close
+    plt.close()  # backend is responsible for flushing any events here
```
It seems this needs to be reverted as it breaks wx testing.

richardsheridan replied:
Am I correct in noticing that this change only produces errors in python 3.8? Maybe that should just be an XFAIL?

richardsheridan replied:
I'll revert and make a TODO comment here to address it in a later PR.

QuLogic replied:
It's hard to say that, as it appears the other versions don't have wx installed. And on the other hand, macOS with Python 3.8 seems to have it, but passes.

richardsheridan replied:
> It seems this needs to be reverted as it breaks wx testing.

I forgot that this change was necessary to get tkagg working after the change.  an alternative could be `plt.pause` or actually going forward with the wx xfail

QuLogic replied:
I did this.

tacaswell commented on lib/matplotlib/_pylab_helpers.py:
```diff
@@ -65,7 +65,8 @@ def destroy(cls, num):
         manager.destroy()
-        gc.collect(1)
+        del manager, num
+        gc.collect()
```
The `del` is OK, but this should go back to being `gc.collect(1)`.  This was an intentional change in https://github.com/matplotlib/matplotlib/pull/3045 where we were running into performance issues where the user had a large number of objects and forcing frequently full cyclic collections was causing them problems. 

For the most part we should trust Python to do memory management and not try to "help"

```suggestion
        # Full cyclic garbage collection may be too expensive to do on every 
        # figure destruction https://github.com/matplotlib/matplotlib/pull/3045
        gc.collect(1)
```

richardsheridan replied:
I firmly agree with not causing a performance regression, good catch. Actually, [this comment](https://github.com/matplotlib/matplotlib/pull/3045#issuecomment-42482288) in that pr corrected my understanding of the function, I thought it was only asking for gen 1 alone to be collected.

QuLogic commented:
I pushed a couple of updates to the tests that should hopefully fix them. At least, they appear fine locally.

richardsheridan commented:
thanks @QuLogic I'm happy with your changes fwiw.

this one is a definite squash merge when the time comes

QuLogic commented:
For reference, using the example from https://github.com/matplotlib/matplotlib/issues/20490#issuecomment-997285267 (+ a `gc.collect`), I managed to run memray on it, and it went from:
![newplot(1)](https://user-images.githubusercontent.com/302469/165845790-1e3e06e1-e859-40f2-a46d-c22bef0736f4.png)
to:
![newplot](https://user-images.githubusercontent.com/302469/165845815-6160b35e-8e99-4d03-851c-e1489e7d520d.png)
There's a bit of a spinup, which I haven't quite investigated (i.e., how many figures, or if it's just initial imports and such). The peaking behaviour doesn't exhibit in the RSS when you run the script by itself, so I assume that's just memray periodically writing out stats (or maybe it's slowed down the gc a bit.)

tacaswell commented:
Thank you @richardsheridan and @QuLogic !

tacaswell commented:
@meeseeksdev backport to v3.5.x
